src/service/auth.rs

```diff
@@ -0,0 +1,122 @@
+const AUTH_METHOD_NAME: &str = "Check";
+
+pub struct AuthService {
+    endpoint: String,
+    metadata: Vec<(String, Vec<u8>)>,
+}
+
+impl AuthService {
+    pub fn new(endpoint: &str, metadata: Vec<(&str, &[u8])>) -> Self {
+        let m = metadata
+            .into_iter()
+            .map(|(header, value)| (header.to_owned(), value.to_owned()))
+            .collect();
+        Self {
+            endpoint: endpoint.to_owned(),
+            metadata: m,
+        }
+    }
+
+    pub fn message() -> CheckRequest {
+        AuthService::build_check_req()
+    }
+
+    fn build_check_req() -> CheckRequest {
+        let mut auth_req = CheckRequest::default();
+        let mut attr = AttributeContext::default();
+        attr.set_request(AuthService::build_request());
+        attr.set_destination(AuthService::build_peer(
+            get_attribute::<String>("destination.address").unwrap_or_default(),
+            get_attribute::<i64>("destination.port").unwrap_or_default() as u32,
+        ));
+        attr.set_source(AuthService::build_peer(
+            get_attribute::<String>("source.address").unwrap_or_default(),
+            get_attribute::<i64>("source.port").unwrap_or_default() as u32,
+        ));
+        // todo(adam-cattermole): for now we set the context_extensions to the request host
+        // but this should take other info into account
+        let context_extensions = HashMap::from([(
+            "host".to_string(),
+            attr.get_request().get_http().host.to_owned(),
+        )]);
+        attr.set_context_extensions(context_extensions);
+        attr.set_metadata_context(Metadata::default());
+        auth_req.set_attributes(attr);
+        auth_req
+    }
+
+    fn build_request() -> AttributeContext_Request {
+        let mut request = AttributeContext_Request::default();
+        let mut http = AttributeContext_HttpRequest::default();
+        let headers: HashMap<String, String> = hostcalls::get_map(MapType::HttpRequestHeaders)
+            .unwrap()
+            .into_iter()
+            .collect();
+
+        http.set_host(get_attribute::<String>("request.host").unwrap_or_default());
+        http.set_method(get_attribute::<String>("request.method").unwrap_or_default());
+        http.set_scheme(get_attribute::<String>("request.scheme").unwrap_or_default());
+        http.set_path(get_attribute::<String>("request.path").unwrap_or_default());
+        http.set_protocol(get_attribute::<String>("request.protocol").unwrap_or_default());
+
+        http.set_headers(headers);
+        request.set_time(get_attribute("request.time").map_or(
+            Timestamp::new(),
+            |date_time: DateTime<FixedOffset>| Timestamp {
+                nanos: date_time.nanosecond() as i32,
+                seconds: date_time.second() as i64,
+                unknown_fields: Default::default(),
+                cached_size: Default::default(),
+            },
+        ));
```

Comment:
I need to investigate how this field is expected to be set as this implementation is incorrect

Reply:
I'm going to iterate on this at a later date